docs/src/main/asciidoc/security-openid-connect-multitenancy-concept.adoc

```diff
@@ -0,0 +1,317 @@
+
+The 3 required steps are detailed below, along with descriptions of concepts and some considerations:
+
+. Find `tenant identifier` from request, and lookup/resolve the corresponding OIDC configuration for the active tenant.
+. Inject user information (packaged in a `JsonWebToken` object) from the active `AccessToken`. This includes implementing reusable logic to simplify access to information given by OIDC.
+. Validate user authorization to retrieve or create data.
+
+=== 1. Resolve the `tenant identifier` and `OidcTenantConfig`
+
+Before being able to execute the actual business logic of a multi-tenant application, Quarkus needs to enforce security requirements for authentication and authorization. To do that, Quarkus must verify that received `JsonWebTokens` are valid, and that the information within them permits the active user to read or update the current resource. In order to do so, Quarkus must read the active `tenant id` of the current request and use it to resolve the active tenant's OIDC configuration (defined by an instance of `OidcTenantConfig`). Finding the tenant id from the current request can be done in one of 2 ways:
+
+. Provide static configuration for all known tenants and the corresponding OIDC configuration for each of them within the `application.properties` file, or
+. Implement a `@Singleton` CDI bean which implements the `io.quarkus.oidc.TenantConfigResolver` interface. The resolver will be invoked for each inbound call to fetch the OIDC configuration for the active tenant id.
+
+The former approach will become quite cumbersome with increasing number of tenants and environments, so if your application is intended to work with more than few tenants it is recommendable to opt for a more scalable, dynamic approach. Also, it is imperative for enterprises maintaining many multi-tenant applications to use _the same algorithm_ for deriving the tenant identifier and OIDC configuration from inbound requests. This yields the necessary usability and maintainability to enable running a plethora of multi-tenant applications. While technically anything within a service request can be used to find the tenant id, it is practical to *embed the tenant id as a resource path segment within the URL*.
+
+[NOTE]
+====
+**_Question:_** Why is it "practical" to embed the tenant ID as a path segment? Why not as a custom HTTP Header or through other means?
+
+**_(Short) Answer:_** To reduce complexities regarding CORS-aware applications, it is practical to embed the tenant id within the path than most other means.
```

Comment:
Good idea to have an FAQ section :-), but I'm concerned having `CORS` in the answer can lead to many more questions :-). Perhaps it can make sense to clarify that custom headers may be considers unsafe with CORS ?

Reply:
Oops - you did clarify it below, sorry :-)